core/thread_interceptor.go

```diff
@@ -0,0 +1,193 @@
+package core
+
+import (
+	"context"
+	"errors"
+
+	ma "github.com/multiformats/go-multiaddr"
+	dbpb "github.com/textileio/go-threads/api/pb"
+	"github.com/textileio/go-threads/core/thread"
+	netpb "github.com/textileio/go-threads/net/api/pb"
+	"github.com/textileio/textile/v2/api/common"
+	mdb "github.com/textileio/textile/v2/mongodb"
+	"go.mongodb.org/mongo-driver/mongo"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// threadInterceptor monitors for thread creation and deletion.
+// Textile tracks threads against dev, org, and user accounts.
+// Users must supply a valid API key from a dev/org.
+func (t *Textile) threadInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		method, _ := grpc.Method(ctx)
+		for _, ignored := range authIgnoredMethods {
+			if method == ignored {
+				return handler(ctx, req)
+			}
+		}
+		for _, block := range blockMethods {
+			if method == block {
+				return nil, status.Error(codes.PermissionDenied, "Method is not accessible")
+			}
+		}
+		if sid, ok := common.SessionFromContext(ctx); ok && sid == t.internalHubSession {
+			return handler(ctx, req)
+		}
+
+		var owner thread.PubKey
+		if org, ok := mdb.OrgFromContext(ctx); ok {
+			owner = org.Key
+		} else if dev, ok := mdb.DevFromContext(ctx); ok {
+			owner = dev.Key
+		} else if user, ok := mdb.UserFromContext(ctx); ok {
+			owner = user.Key
+		}
+
+		var newID thread.ID
+		var isDB bool
+		var err error
+		switch method {
+		case "/threads.pb.API/NewDB":
```

Comment:
Unrleated of this PR, idea: we rely on gRPC APIs as strings. If we somewhat rename a gRPC identifier, many of our things will fail silently. Not only here, but other files and also other of our projects.

I'm not aware if there's a dynamic way to get the name using the generated code of the client.

Reply:
Yep, pretty fragile. Curious if there's a best practice for this kind of method matching. Will look around.